python-package/lightgbm/callback.py

```diff
@@ -229,7 +229,12 @@ def __call__(self, env: CallbackEnv) -> None:
             if new_param != env.params.get(key, None):
                 new_parameters[key] = new_param
         if new_parameters:
-            env.model.reset_parameter(new_parameters)
+            if isinstance(env.model, Booster):
+                env.model.reset_parameter(new_parameters)
+            else:
+                # CVBooster holds a list of Booster objects, each needs to be updated
+                for i in range(len(env.model.boosters)):
+                    env.model.boosters[i].reset_parameter(new_parameters)
```

Comment:
Are you sure that'll work?

I specifically chose to iteration over positions to ensure I was modifying the existing Boosters in the `CVBooster.boosters` *by reference*.

This proposes change is equivalent to the following, for the first iteration:

```python
bst = env.model.boosters[0]
bst.reset_parameter(new_parameters)
```

Would that actually update the object at `env.model.boosters[0]` in place? I think it might not and it'd instead be creating and modifying a new object.

If you feel very strongly that using positional indexing is undesirable, I'll spend the time to test this. If this is just a minor stylistic difference, I'd prefer to leave it as-is.

Reply:
Yes, that works because the method modifies the object in-place, so in each iteration we get the object and modify it. But it's fine to leave it as it is if you want.

Reply:
> because the method modifies the object in-place

I understand that, but `for booster in boosters` also allocates a new variable `booster` and assigns each `Booster` to it as it iterates over the list.

I thought modifying that new variable might cause a copy to be created, but I guess not.

Just to show clearly what I'm talking about:

```python
import lightgbm as lgb
from sklearn.datasets import make_regression

X, y = make_regression(n_samples=10_000)
dtrain = lgb.Dataset(X, label=y)

results = lgb.cv(
    train_set=dtrain,
    params={"objective": "regression"},
    num_boost_round=7,
    nfold=3,
    stratified=False,
    return_cvbooster=True
)

cv_bst = results["cvbooster"]

cv_bst.boosters[0].params
# {'objective': 'regression', 'num_iterations': 7}

# list objects in this scope
[x for x in dir() if not x.startswith("_")]
# ['X', 'cv_bst', 'dtrain', 'lgb', 'make_regression', 'results', 'y']

for booster in cv_bst.boosters:
    booster.params = {"objective": "changed"}

# there's no a new object 'booster' in this scope
[x for x in dir() if not x.startswith("_")]
# ['X', 'booster', 'cv_bst', 'dtrain', 'lgb', 'make_regression', 'results', 'y']

# ... but looks like it must just be a reference to the one on CVBooster
#     (the last one, specifically)
assert id(booster) == id(cv_bst.boosters[-1])

# ... and you can see that by indexing into `CVBooster.boosters`
cv_bst.boosters[0].params
# {'objective': 'changed'}
```

This was a perfect mix of my paranoia and ignorance of when Python does things by reference vs. by value 😅 .

I'll accept your suggestion, thank you.